### ISSUE TITLE
reconnect helper for change streams

### ISSUE THREAD

**Issue by vkarpov15** [Automattic/mongoose#6799]
Re: #6741, change streams currently fail when there's a replica set failover. We should have some mechanism in place to handle this case. @mbroadst what are your thoughts?

**Comment by mbroadst** [comment]
@vkarpov15 change streams were designed explicitly to support handling failover, see: https://github.com/mongodb/node-mongodb-native/blob/master/lib/change_stream.js#L348-L379. Perhaps you are running into a non-resumable error case? We would need more information about the specific error scenario, and ideally a reproducible case would help us resolve this very quickly.

**Comment by vkarpov15** [comment]
@mbroadst I can repro this consistently in mongoose by running `rs.stepDown()` in the shell while the below script is running:

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh6741`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString, { useNewUrlParser: true, replicaSet: 'rs' });
  await mongoose.connection.dropDatabase();

  await mongoose.connection.createCollection('Foo');
  
  const M = mongoose.model('Foo', new Schema({ name: String }), 'Foo');
  
  let changeStream = M.watch().on('change', change => console.log(change));
  console.log('Watching...');
}
```

Do you need a driver-only script or is this enough to get you started?

**Comment by mbroadst** [comment]
@vkarpov15 that's enough for me, and yeah it looks like there's a bug here. Our resume logic is only presently tested when using the change stream as a cursor, not as a stream. I made [NODE-1617](https://jira.mongodb.org/browse/NODE-1617) to track this, we'll get on a fix ASAP.

**Comment by vkarpov15** [comment]
Should be fixed by #6920
